app/celery/celery.py

```diff
@@ -73,3 +84,13 @@ def init_app(self, app):
         )
 
         self.conf.update(app.config['CELERY'])
+
+    def send_task(self, name, args=None, kwargs=None, **other_kwargs):
+        kwargs = kwargs or {}
+
+        if has_request_context() and hasattr(request, 'request_id'):
+            kwargs['request_id'] = request.request_id
+        elif has_app_context() and 'request_id' in g:
+            kwargs['request_id'] = g.request_id
+
+        return super().send_task(name, args, kwargs, **other_kwargs)
```

Comment:
this is a great point adding this, and one I only thought about this morning when looking at ftp - the antivirus app won't receive any request IDs because api won't call them when using `send_task` to invoke the antivirus app (https://github.com/alphagov/notifications-api/blob/b0c531f5cb5b9df1cb1ee79e933d97ccdea65a34/app/v2/notifications/post_notifications.py#L434-L438).

Looking at celery source code, it looks like `apply_async` calls `send_task` internally anyway, (https://github.com/celery/celery/blob/4fad9072ff4eca154ab0ac0b76f3a54fd7e738fe/celery/app/task.py#L561). I think we might be able to remove the overriden apply_async completely - this `NotifyCelery.send_task` override should take care of everything. Obviously some testing required.

Reply:
> the antivirus app won't receive any request IDs

This is intentional for now: we need to add support for the `request_id` kwarg before sending it, otherwise the task itself will get it unexpectedly and error. I forgot about FTP, though, so I'll try and do an equivalent PR for that once this one and Template Preview are in, so we can actually get the whole chain working.

> Looking at celery source code, it looks like apply_async calls send_task internally anyway

Great spot! I've now done some local testing and can confirm a fake request ID goes all the way through API => Antivirus => Template Preview => API if I replace `apply_async -> send_task` in the API (temporarily). I've added an extra commit to this PR and the Template Preview PR to explain why we don't need it.

I think this is ready for another look when you have a moment, and maybe https://github.com/alphagov/notifications-template-preview/pull/531 as well, since it's basically the same.